### PR TITLE
Price Watch Interval Setting

### LINKED ISSUE #25: Custom Price Watch Interval feature
Hello, 

I would like to point out one cool feature, that would allow people to set their own interval for `price-watch-job` and `price-watch-single-job`.

For instance, I would like to have the stats posted every 24 hours, and the price every hour, as I would like to avoid the bot spamming on the channel. Right now, I saw that it's hardcoded.

Probably, the bot would need two additional commands for the Administrator.  `!statsInterval` and `!priceInterval`.

If you accept this feature, I would be happy to work on it and create a Pull Request, as it would be my first contribution to the open source project.

### PR DESCRIPTION
**Pull request by ernest-rudnicki**
### Description
The administrators should have possibility to change time intervals, as some of the jobs that have small intervals, like for instance `price-watch-single` can be annoying, and they can cause a spam on channels.

This pull request implements functionality to change and adjust time intervals between jobs for guilds administrators. Additionally, it implements a command for checking current time interval. The implementation was done for sharding and single bots. 

### Introduces
- Jobs will now be fulfilled according to the interval set for the specific guild.
- Introducing cache for `price-watch` job.
- Reducing the interval of calling the jobs for `price-watch-single` to 5s, and `price-watch` to 1min.
- `!pricewatch` command will now save two new jobs (price-watch and price-watch-single) for the guild in the database.
- `!jobstart`  accepts also an interval in the format of `0d, 0h, 0m, 0s`. If interval is not provided, it saves a job with a default one.
- `!jobinterval` is a new command that allows to change the interval for arleady started jobs. It accepts same interval format like `!jobstart`.
- `!checkinterval` is a new command that allows to check what current interval is set on the existing job.

### Closes issues
Closes #25

### CODE REVIEW

File: src/commands/admin/jobs/stop.js [hunk 1]
```diff
@@ -38,11 +39,13 @@ module.exports = ({
       }, {
         guildId: guild.id,
         jobName: job.meta.name,
-        jobState: false
+        jobState: false,
+        jobInterval: null,
```
Comment (reecebenson) [inline src/commands/admin/jobs/stop.js]: We probably don't want to reset the guild owners set interval here, we could probably just leave it as it is instead of resetting it.

File: src/commands/admin/jobs/check-interval.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        return await message.lineReply('Please specify a Job name: \n```\n' + Array.from(client.jobs.keys()).join('\n') + '\n```')
+      }
+  
+      if (!Array.from(client.jobs.keys()).includes(jobName)) {
+        return await message.lineReply('Invalid Job. Valid Jobs: \n```\n' + Array.from(client.jobs.keys()).join('\n') + '\n```')
+      }
+
+      const job = client.jobs.get(jobName)
+      if (!job.meta.guildControlled) {
+        return await message.lineReply('Guild controlled jobs run on an interval that cannot be changed.')
```
Comment (reecebenson) [inline src/commands/admin/jobs/check-interval.js]: This is probably a typo from me, but could this read:
`This job is not guild controlled so the interval cannot be changed.`

File: src/commands/admin/jobs/interval.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return await message.lineReply('Please specify a Job name: \n```\n' + Array.from(client.jobs.keys()).join('\n') + '\n```')
+      }
+  
+      if (!Array.from(client.jobs.keys()).includes(jobName)) {
+        return await message.lineReply('Invalid Job. Valid Jobs: \n```\n' + Array.from(client.jobs.keys()).join('\n') + '\n```')
+      }
+
+      const job = client.jobs.get(jobName)
+      if (!job.meta.guildControlled) {
+        return await message.lineReply('Guild controlled jobs run on an interval that cannot be changed.')
```
Comment (reecebenson) [inline src/commands/admin/jobs/interval.js]: This is probably a typo from me, but could this read:
`This job is not guild controlled so the interval cannot be changed.`

File: src/commands/admin/jobs/jobs.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ module.exports = ({
         const status = (entry ? entry.jobState : job.meta.enabled) ? ':white_check_mark:' : ':x:'
 
         return {
-          "name": `${status} - \`${job.meta.name}\` | ${secsToDHMS(job.meta.interval / 1000)}`,
+          "name": `${status} - \`${job.meta.name}\` | ${secsToDHMS(job.meta.defaultInterval ? job.meta.defaultInterval / 1000 : job.meta.interval / 1000)}`,
```
Comment (reecebenson) [inline src/commands/admin/jobs/jobs.js]: The logic here seems incorrect. We should be defaulting to `job.meta.defaultInterval` if the row in the database (`jobSchema`, on the `entry` variable) does not exist. I'd expect the following (not tested):

```js
const entry = jobStatus.find((s) => s.jobName == job.meta.name)
const status = (entry ? entry.jobState : job.meta.enabled) ? ':white_check_mark:' : ':x:'
// below checks to see if 'entry' exists, if so, check if jobInterval is null, if not use the interval from db row, otherwise use default
const interval = (entry ? (entry.jobInterval ? entry.jobInterval : job.meta.defaultInterval) : job.meta.defaultInterval

return {
  "name": `${status} - \`${job.meta.name}\` | ${secsToDHMS(interval / 1000)}`,
  ...
}
```

File: src/commands/admin/jobs/interval.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      }
+
+      if(!intervalText) {
```
Comment (reecebenson) [inline src/commands/admin/jobs/interval.js]: linting - space after `if`

File: src/commands/admin/jobs/interval.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+module.exports = ({
+  meta: {
+    name: 'jobs-interval',
+    description: 'Change the interval for job',
+    commands: ['jobinterval'],
+    permissions: ['ADMINISTRATOR']
+  },
+  run: async (client, cache, message) => {
+      const { args } = processCmd(message)
```
Comment (reecebenson) [inline src/commands/admin/jobs/interval.js]: Please remove 2 spaces (unindent) to match the rest of the codebase's style (for the rest of this file).

File: src/commands/admin/jobs/interval.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        }, {
+          jobInterval: interval,
+          lastJobTime: lastJob,
+        }, {
+          upsert: true
+        })
+        return await message.reply(`Successfully changed the interval for ${getInlineCodeblock(jobName)}, this will execute on the jobs next interval.`)
+      }
+      catch {}
+
```
Comment (reecebenson) [inline src/commands/admin/jobs/interval.js]: Unnecessary newline (nitpicky)

File: src/commands/admin/jobs/start.js [hunk 8]
```diff
@@ -24,6 +30,30 @@ module.exports = ({
...
+        return await message.lineReply(`The interval cannot be lesser or equal to ${secsToDHMS(Math.floor(job.meta.defaultInterval / 1000))}.`)
+      }
+
+      interval = Math.floor(interval / 1000)
+      lastJob = getISODate()
+
+    } else if (job.meta.defaultInterval) {
+      interval = Math.floor(job.meta.defaultInterval / 1000)
+      lastJob = getISODate()
+    }
```
Comment (reecebenson) [inline src/commands/admin/jobs/start.js:55]: Should we have an `else` here for catching an error if `intervalText` is null / empty?

File: src/jobs/pricewatch.js [hunk 9]
```diff
@@ -1,25 +1,47 @@
...
+const { timeNow, getISODate } = require('../utils/helper');
+const { getExcludedGuilds } = require('../utils/jobs')
+const { DEFAULT_PRICE_WATCH_INTERVAL } = require('../constants/constants')
 
 module.exports = ({
   meta: {
     name: 'price-watch',
     description: 'Sends an embed message of the current statistics of SafeMoon',
-    interval: 300 * 1000,
+    interval: 60 * 1000,
```
Comment (reecebenson) [inline src/jobs/pricewatch.js:12]: This `interval` should never be less than the `cacheTime` variable, otherwise it'll just output the same data each time.

File: src/jobs/pricewatch-single.js [hunk 10]
```diff
@@ -89,17 +96,29 @@ module.exports = ({
...
+            await jobSchema.updateMany(
+              { 
+                'guildId': { $in : watchingGuilds },
+                'jobName': { $eq: jobName }
+              },
+              { $set: { 'lastJobTime': getISODate() } }
+            )
+          } catch {}
         })
+
```
Comment (reecebenson) [inline src/jobs/pricewatch-single.js]: Unnecessary newline (nitpicky)

File: src/jobs/pricewatch-single.js [hunk 11]
```diff
@@ -72,6 +78,7 @@ module.exports = ({
               })
 
               watchingGuilds.forEach(async (guildId) => {
+
```
Comment (reecebenson) [inline src/jobs/pricewatch-single.js]: Unnecessary newline (nitpicky)

File: src/jobs/pricewatch-single.js [hunk 12]
```diff
@@ -45,15 +49,17 @@ module.exports = ({
...
+        const excludedGuilds = getExcludedGuilds(jobs)
 
         // Find guilds associated to each shard
         const shardedGuildIds = await client.shard.broadcastEval(`this.guilds.cache.map((g) => g.id)`)
         shardedGuildIds.forEach(async (guildIds, shardId) => {
+
           const watchingGuilds = guildIds.filter((dbId) => !excludedGuilds.includes(dbId))
           const guildsByCsv = watchingGuilds.join(',')
           const dbGuildsAndChannels = guilds.map((g) => `${g.id}/${g.channelId}`).join(',')
+
```
Comment (reecebenson) [inline src/jobs/pricewatch-single.js]: Unnecessary newline (nitpicky)

File: src/jobs/pricewatch-single.js [hunk 13]
```diff
@@ -1,13 +1,16 @@
...
+const { timeNow, getISODate } = require('../utils/helper')
+const { getExcludedGuilds } = require('../utils/jobs')
+const { DEFAULT_PRICE_WATCH_SINGLE_INTERVAL } = require('../constants/constants')
 
 module.exports = ({
   meta: {
     name: 'price-watch-single',
     description: 'Sends a single message of the current price of SafeMoon',
-    interval: 15 * 1000,
+    interval: 5 * 1000,
```
Comment (reecebenson) [inline src/jobs/pricewatch-single.js:12]: This `interval` should never be less than the `cacheTime` variable, otherwise it'll just output the same data each time, and preferably never below 15 to avoid API abuse.

File: src/jobs/pricewatch.js [hunk 14]
```diff
@@ -60,12 +82,24 @@ module.exports = ({
...
+
+          try {
+            await jobSchema.updateMany(
+              { 
+                'guildId': { $in : watchingGuilds },
+                'jobName': { $eq: jobName }
+              },
+              { $set: {'lastJobTime': getISODate() } }
+            )
+
```
Comment (reecebenson) [inline src/jobs/pricewatch.js]: Unnecessary newline (nitpicky)

File: src/jobs/pricewatch.js [hunk 15]
```diff
@@ -60,12 +82,24 @@ module.exports = ({
...
+            await jobSchema.updateMany(
+              { 
+                'guildId': { $in : watchingGuilds },
+                'jobName': { $eq: jobName }
+              },
+              { $set: {'lastJobTime': getISODate() } }
+            )
+            
+          } catch {}
+
```
Comment (reecebenson) [inline src/jobs/pricewatch.js]: Unnecessary newline (nitpicky)

File: src/jobs/pricewatch.js [hunk 16]
```diff
@@ -60,12 +82,24 @@ module.exports = ({
...
+                'jobName': { $eq: jobName }
+              },
+              { $set: {'lastJobTime': getISODate() } }
+            )
+            
+          } catch {}
+          
         })
-      }
+        }
```
Comment (reecebenson) [inline src/jobs/pricewatch.js]: Invalid indentation level (-2sp)

File: src/jobs/roleschedule.js [hunk 17]
```diff
@@ -17,9 +18,9 @@ module.exports = ({
       // Are we sharding?
       if (client.shard) {
         // Find guilds that are excluded from this job
-        const jobsExcluded = await jobSchema.find({ jobState: false, jobName: 'role-schedule' })
-        const excludedGuilds = JSON.stringify(jobsExcluded.map((exc) => exc.guildId))
-
+        const jobs = await jobSchema.find({ jobName: 'role-schedule' })
+        const excludedGuilds = JSON.stringify(getExcludedGuilds(jobs))
+
```
Comment (reecebenson) [inline src/jobs/roleschedule.js]: Unnecessary whitespace (nitpicky)

File: src/utils/jobs.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+    if (
+      job.jobInterval && 
+      !didTimePassed(job.lastJobTime, job.jobInterval)
+      ) {
+      excGuilds.push(job.guildId)
+    }
+  })
+  return excGuilds
+}
```
Comment (reecebenson) [inline src/utils/jobs.js]: This could be condensed for ES6:
```js
const getExcludedGuilds = (jobs) => {
  return jobs.filter(
    (job) => !job.jobState || (job.jobInterval && !didTimePassed(job.lastJobTime, job.jobInterval))
  )
}
```

**Review by reecebenson — CHANGES_REQUESTED** [review]

File: src/commands/admin/jobs/jobs.js (hunk 4 above)
Comment (Terminats) [reply src/commands/admin/jobs/jobs.js]: Yes, however I didn't set `defaultInterval` for all of the jobs, since the ones not controlled by the guild would have an ` interval` and `defaultInterval` set to same values. Here's needed an additional check, whether ` interval` is `undefined`, if so, the value from `job.meta.interval` should be used.

File: src/utils/jobs.js (hunk 18 above)
Comment (Terminats) [reply src/utils/jobs.js]: Not quite actually, since this function should return only the ids of the guilds, and not whole objects. But I will change the name, because it is not clear enough what it does.

File: src/commands/admin/jobs/start.js (hunk 8 above)
Comment (Terminats) [reply src/commands/admin/jobs/start.js:55]: I don't think so, if `intervalText` is empty, then it will just use `defaultInterval`, if there is no `defaultInterval`, then the job will be run on the `interval` that runs the code.

File: src/jobs/pricewatch-single.js (hunk 13 above)
Comment (Terminats) [reply src/jobs/pricewatch-single.js:12]: I decreased it on purpose, so the difference between the passed time of certain guilds is not that high.

Let's say one guild has the time interval set to 15 seconds. The job run just one second before the interval passed for this guild. If that happens, this guild has to wait additional 15 seconds, before it will have the job executed. That makes 29 seconds for the job to be executed for that guild. However, if we set it just to 5 seconds. That guild would have to wait only additional 6 seconds for the job to be executed.

The default interval is responsible here for not allowing to post the same data. So if the defaultInterval is set to 15 seconds and the user is not allowed to set own interval below that value, the guild should not have the values repeated.

File: src/jobs/pricewatch.js (hunk 9 above)
Comment (Terminats) [reply src/jobs/pricewatch.js:12]: Same as above in the commnet

File: src/utils/jobs.js (hunk 18 above)
Comment (reecebenson) [reply src/utils/jobs.js]: Oh, woops - missed that. You can then use:

```js
const getExcludedGuilds = (jobs) => {
  return jobs
    .filter((job) => !job.jobState || (job.jobInterval && !didTimePassed(job.lastJobTime, job.jobInterval)))
    .map((job) => job.guildId)
}
```

File: src/utils/jobs.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
+const { didTimePassed } = require("./helper")
+
+const getExcludedGuildsIds = (jobs) => {
+  const excGuilds = []
+  jobs.forEach((job) => {
+    if (!job.jobState || (job.jobInterval && !didTimePassed(job.lastJobTime, job.jobInterval))) {
+      excGuilds.push(job.guildId)
+    }
+  })
+  return excGuilds
```
Comment (reecebenson) [inline src/utils/jobs.js]: As per your reply, we should only have Guild IDs here, but we can still use ES6 logic to map:
```js
const getExcludedGuildsIds = (jobs) => {
  return jobs
    .filter((job) => !job.jobState || (job.jobInterval && !didTimePassed(job.lastJobTime, job.jobInterval)))
    .map((job) => job.guildId)
}
```

**Review by reecebenson — CHANGES_REQUESTED** [review]
One final change - everything else looks good!

**Review by reecebenson — APPROVED** [review]

File: src/utils/jobs.js (hunk 19 above)
Comment (vitaliiburlaka) [reply src/utils/jobs.js]: You could even do all that in a single loop using just `.map()`
```js
const getExcludedGuildsIds = (jobs) => {
  return jobs
    .map((job) => {
       if (!job.jobState || (job.jobInterval && !didTimePassed(job.lastJobTime, job.jobInterval))) {
         return job.guildId
       }
    })
}
```

Comment (reecebenson) [reply src/utils/jobs.js]: That's more lines 😄
